### PR TITLE
CORTX-32043: RGW Error response handled

### PR DESCRIPTION
**Pull request by abhijit1patil**
Signed-off-by: abhijit1patil <abhijit.1.patil@seagate.com>

# Pull Request
## Problem Statement
-  [CORTX-32374](https://jts.seagate.com/browse/CORTX-32374)

## Design
-   RGW Error Code added

## Custom Build and Deployment

- [Custom Build Job](https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/7085/)
- Custom Build
     cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-7085-custom-ci
     cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-7085-custom-ci
     cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-7085-custom-ci
- [Jenkins Jobs](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/8474/)

-   \[ \] Coding conventions are followed and code is consistent

## Dev Testing
1. If user owns a resource - 
image

2. If CSM is unable to parse error code thrown by rgw then unknown error will be returned, earlier code used to break due to keyerror.
image

-   \[ \] Unit and System Tests are added
-   \[ \] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   \[ \] Testing was performed with RPM

-   \[ \] Interface change (if any) are documented
-   \[ \] Side effects on other features (deployment/upgrade)
-   \[ \] Dependencies on other component(s)

-   \[ \] JIRA number/GitHub Issue added to PR
-   \[ \] PR is self reviewed
-   \[ \] Jira and state/status is updated and JIRA is updated with PR link
-   \[ \] Check if the description is clear and explained

-   \[ \] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/plugins/cortx/rgw.py [hunk 1]
```diff
@@ -150,7 +150,12 @@ def _create_error(self, status: int, body: dict) -> Any:
...
         rgw_error.http_status = status
-        rgw_error.error_code = RgwErrors[body['Code']]
-        rgw_error.error_message = rgw_error.error_code.value
-
+        try:
+            rgw_error.error_code = RgwErrors[body['Code']]
+            rgw_error.error_message = rgw_error.error_code.value
+        except ValueError:
+            Log.error("Unable to parse error code received from S3,"
+                      "Generating generic error response.")
```
Comment (rohitkolapkar) [inline csm/plugins/cortx/rgw.py:158]: Please use one indentation rule.

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/plugins/cortx/rgw.py (hunk 1 above)
Comment (abhijit1patil) [reply csm/plugins/cortx/rgw.py]: updated

File: csm/core/data/models/rgw.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ class RgwErrors(Enum):
     EmailExists = "Provided email address exists."
     InvalidCapability = "Attempt to grant invalid admin capability."
     InvalidTenantName = "Invalid tenant name provided."
+    BucketAlreadyExists = "Bucket resource exists."
```
Comment (pawankumarsrivastava) [inline csm/core/data/models/rgw.py:48]: This does not look good. Need to discuss. The message should clearly convey that unable to delete IAM user.

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/data/models/rgw.py (hunk 2 above)
Comment (rohitkolapkar) [reply csm/core/data/models/rgw.py]: updated

**Review by rohitkolapkar — APPROVED** [review]

File: csm/plugins/cortx/rgw.py [hunk 3]
```diff
@@ -149,8 +149,14 @@ def _create_error(self, status: int, body: dict) -> Any:
...
         rgw_error = RgwError()
-        rgw_error.http_status = status
-        rgw_error.error_code = RgwErrors[body['Code']]
-        rgw_error.error_message = rgw_error.error_code.value
-
+        try:
+            rgw_error.http_status = status
+            rgw_error.error_code = RgwErrors[body['Code']]
+            rgw_error.error_message = rgw_error.error_code.value
+        except KeyError:
```
Comment (pawankumarsrivastava) [inline csm/plugins/cortx/rgw.py:156]: What if other than KeyError is raised due to any unknown issue? Should we handle base exception too?

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/plugins/cortx/rgw.py (hunk 3 above)
Comment (rohitkolapkar) [reply csm/plugins/cortx/rgw.py:156]: The only possible error at this stage in the code flow is 'KeyError'.

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**